### PR TITLE
add run_if_changed/RunsAgainstChanges for prow postsubmit jobs

### PR DESCRIPTION
**Pull request by ibzib**
Create a `ChangeMatcher` struct to enable `RunsAgainstChanges` on prow postsubmit jobs (this was already present in presubmits, but not shared). This will allow Prow components (such as the Gerrit adapter) to selectively run postsubmit jobs depending on which files were changed.

/shrug
/assign @krzyzacy 
/cc @BenTheElder

### CODE REVIEW

**Comment by krzyzacy** [comment]
what about the place postsubmit is actually triggered?

also
/cc @cjwagner

File: prow/config/jobs.go [hunk 1]
```diff
@@ -145,6 +145,10 @@ type Presubmit struct {
 type Postsubmit struct {
 	JobBase
 
+	// RunIfChanged automatically run if the PR modifies a file that matches this regex.
```
Comment (rkpagadala) [inline prow/config/jobs.go]: Should these changes go into JobBase itself? since both Postsubmit and Presubmit use it?

Comment (krzyzacy) [reply prow/config/jobs.go]: periodic doesn't use it though

Comment (rkpagadala) [reply prow/config/jobs.go]: ah.. ok

Comment (ibzib) [reply prow/config/jobs.go]: How about a struct similar to [`Brancher`](https://github.com/kubernetes/test-infra/blob/3c763287157e7040f51d3d84fec2e1e433e9d4df/prow/config/jobs.go#L180-L191) to contain the shared logic?

Comment (rkpagadala) [reply prow/config/jobs.go]: +1

File: prow/config/jobs.go [hunk 2]
```diff
@@ -237,16 +241,25 @@ func (br Brancher) Intersects(other Brancher) bool {
...
 	return false
 }
 
+// RunsAgainstChanges returns true if any of the changed input paths match the run_if_changed regex.
+func (ps Presubmit) RunsAgainstChanges(changes []string) bool {
+	return matchChanges(ps.reChanges, changes)
+}
+
+// RunsAgainstChanges returns true if any of the changed input paths match the run_if_changed regex.
+func (ps Postsubmit) RunsAgainstChanges(changes []string) bool {
```
[the remark this answers is not in the document]
Comment (ibzib) [reply prow/config/jobs.go]: Looks like it's only needed in `push.go`. I'll use `RunsAgainstChanges` in the Gerrit adapter in my next PR.

File: prow/config/jobs.go [hunk 3]
```diff
@@ -190,6 +191,13 @@ type Brancher struct {
 	reSkip *regexp.Regexp
 }
 
+// ChangeMatcher is for code shared between jobs that run only when certain files are changed.
+type ChangeMatcher struct {
+	// RunIfChanged automatically run if the PR modifies a file that matches this regex.
```
Comment (stevekuznetsov) [inline prow/config/jobs.go]:
```
// RunIfChanged defines a regex used to select which subset of file changes should trigger this job.
// If any file in the changeset matches this regex, the job will be triggered
```

or similar?

File: prow/config/jobs.go [hunk 4]
```diff
@@ -190,6 +191,13 @@ type Brancher struct {
...
+	RunIfChanged string         `json:"run_if_changed,omitempty"`
+	reChanges    *regexp.Regexp // from RunIfChanged
```
Comment (stevekuznetsov) [inline prow/config/jobs.go]: This name will be unwieldy if there's ever another type of `ChangeMatcher` backing impl in this struct

File: prow/config/jobs.go [hunk 5]
```diff
@@ -238,9 +246,12 @@ func (br Brancher) Intersects(other Brancher) bool {
 }
 
 // RunsAgainstChanges returns true if any of the changed input paths match the run_if_changed regex.
-func (ps Presubmit) RunsAgainstChanges(changes []string) bool {
+func (cm ChangeMatcher) RunsAgainstChanges(changes []string) bool {
+	if cm.RunIfChanged == "" {
```
Comment (stevekuznetsov) [inline prow/config/jobs.go]: Why didn't this check need to exist before?

Comment (ibzib) [reply prow/config/jobs.go]: It seems the check was already being made everywhere `RunsAgainstChanges` was used. `reChanges` [isn't always set](https://github.com/kubernetes/test-infra/blob/3c763287157e7040f51d3d84fec2e1e433e9d4df/prow/config/config.go#L1172), creating the risk of nil deref, so I figured it'd be safer to put the check in the function here.

**Comment by stevekuznetsov** [comment]
/lgtm
/hold
@krzyzacy

**Comment by krzyzacy** [comment]
feel free to `/hold cancel`

**Pull request merged**